src/Aspire.Hosting/Eventing/DistributedApplicationEventing.cs

```diff
@@ -2,36 +2,78 @@
 
 /// <inheritdoc cref="IDistributedApplicationEventing" />
-[Experimental("ASPIREEVENTING001", UrlFormat = "https://aka.ms/dotnet/aspire/diagnostics#{0}")]
 public class DistributedApplicationEventing : IDistributedApplicationEventing
 {
     private readonly ConcurrentDictionary<Type, List<DistributedApplicationEventSubscription>> _eventSubscriptionListLookup = new();
     private readonly ConcurrentDictionary<DistributedApplicationEventSubscription, Type> _subscriptionEventTypeLookup = new();
 
     /// <inheritdoc cref="IDistributedApplicationEventing.PublishAsync{T}(T, CancellationToken)" />
-    [Experimental("ASPIREEVENTING001", UrlFormat = "https://aka.ms/dotnet/aspire/diagnostics#{0}")]
-    public async Task PublishAsync<T>(T @event, CancellationToken cancellationToken = default) where T : IDistributedApplicationEvent
+    [System.Diagnostics.CodeAnalysis.SuppressMessage("ApiDesign", "RS0026:Do not add multiple public overloads with optional parameters", Justification = "Cancellation token")]
+    public Task PublishAsync<T>(T @event, CancellationToken cancellationToken = default) where T : IDistributedApplicationEvent
+    {
+        return PublishAsync(@event, EventDispatchBehavior.BlockingSequential, cancellationToken);
+    }
+
+    /// <inheritdoc cref="IDistributedApplicationEventing.PublishAsync{T}(T, CancellationToken)" />
+    [System.Diagnostics.CodeAnalysis.SuppressMessage("ApiDesign", "RS0026:Do not add multiple public overloads with optional parameters", Justification = "Cancellation token")]
+    public async Task PublishAsync<T>(T @event, EventDispatchBehavior dispatchBehavior, CancellationToken cancellationToken = default) where T : IDistributedApplicationEvent
     {
         if (_eventSubscriptionListLookup.TryGetValue(typeof(T), out var subscriptions))
         {
-            // Taking a snapshot of the subscription list to avoid any concurrency issues
-            // whilst we iterate over the subscriptions. Subscribers could result in the
-            // subscriptions being removed from the list.
-            foreach (var subscription in subscriptions.ToArray())
+            if (dispatchBehavior == EventDispatchBehavior.BlockingConcurrent || dispatchBehavior == EventDispatchBehavior.NonBlockingConcurrent)
+            {
+                var pendingSubscriptionCallbacks = new List<Task>(subscriptions.Count);
+                foreach (var subscription in subscriptions.ToArray())
+                {
+                    var pendingSubscriptionCallback = subscription.Callback(@event, cancellationToken);
+                    pendingSubscriptionCallbacks.Add(pendingSubscriptionCallback);
+                }
+
+                if (dispatchBehavior == EventDispatchBehavior.NonBlockingConcurrent)
+                {
+                    // Non-blocking concurrent.
+                    _ = Task.Run(async () =>
+                    {
+                        await Task.WhenAll(pendingSubscriptionCallbacks).ConfigureAwait(false);
```

Comment:
Log failures here

Reply:
We can't :( Logging isn't necessarily setup yet.